File: frontend/src/lib/menuRuleTemplates.ts
import type { MenuRule } from "@/types"
import { PROTEIN_OPTIONS } from "@/lib/weekPlanCombos"

// Mirrors the backend MenuRuleTemplate JSON schema; the backend re-renders
// the executable expression from this on save/validate.
export interface MenuRuleTemplate {
  type: "limit" | "no_repeat" | "prefer" | "avoid"
  scope?: "meal" | "day" | "week"
  category?: string
  n?: number
  points?: number
  strength?: "must" | "prefer"
}

export const TYPE_OPTIONS: Array<{ key: MenuRuleTemplate["type"]; label: string }> = [
  { key: "limit", label: "最多出现" },
  { key: "no_repeat", label: "主蛋白不重样" },
  { key: "prefer", label: "优先安排" },
  { key: "avoid", label: "尽量避开" },
]

export const SCOPE_OPTIONS: Array<{ key: "meal" | "day" | "week"; label: string }> = [
  { key: "meal", label: "同一餐" },
  { key: "day", label: "同一天" },
  { key: "week", label: "一周" },
]

export const STRENGTH_OPTIONS: Array<{ key: "must" | "prefer"; label: string }> = [
  { key: "prefer", label: "尽量做到" },
  { key: "must", label: "必须遵守" },
]

const FIXED_CATEGORY_LABELS: Record<string, string> = {
  egg: "蛋类菜",
  cold: "凉菜",
  staple: "主食",
  soup: "汤",
  spicy: "辣菜",
  heavy_spicy: "重油重辣的菜",
  deep_fry: "油炸菜",
  slow: "费时菜（超过45分钟）",
}

// Display-only categories used by the v3/v4 preset rules. Not offered in the
// sentence builder — they exist for rendering the factory rules' sentences.
const DISPLAY_ONLY_CATEGORY_LABELS: Record<string, string> = {
  non_favorite: "没收藏的菜",
  unfamiliar_category: "一道收藏都没有的菜系",
  weekday_slow_soup: "工作日的费时汤",
  slow_soup: "费时汤（炖煮超45分钟）",
  soup_ingredient_repeat: "和昨天的汤撞主料的汤",
}

export const CATEGORY_OPTIONS: Array<{ key: string; label: string }> = [
  ...Object.entries(FIXED_CATEGORY_LABELS).map(([key, label]) => ({ key, label })),
  ...PROTEIN_OPTIONS.map((option) => ({ key: `protein:${option.key}`, label: `${option.label}菜` })),
]

export function categoryLabel(category?: string): string {
  if (!category) return "某类菜"
  if (category.startsWith("protein:")) {
    const key = category.slice("protein:".length)
    const label = PROTEIN_OPTIONS.find((option) => option.key === key)?.label || key
    return `${label}菜`
  }
  if (category.startsWith("ingredient:")) {
    return `带「${category.slice("ingredient:".length)}」的菜`
  }
  return FIXED_CATEGORY_LABELS[category] || DISPLAY_ONLY_CATEGORY_LABELS[category] || category
}

const SCOPE_LABELS: Record<string, string> = { meal: "同一餐", day: "同一天", week: "一周" }

export function describeTemplate(tpl: MenuRuleTemplate): string {
  const scope = SCOPE_LABELS[tpl.scope || "meal"]
  const strength = tpl.strength === "must" ? "必须遵守" : "尽量做到"
  const n = tpl.n && tpl.n > 0 ? tpl.n : 1
  switch (tpl.type) {
    case "limit":
      return `${scope}里，${categoryLabel(tpl.category)}最多 ${n} 道 · ${strength}`
    case "no_repeat":
      if ((tpl.scope || "meal") === "meal" && n <= 1) {
        return `同一餐里，主蛋白尽量不重样（豆制品除外） · ${strength}`
      }
      return `${scope}里，同一种主蛋白最多安排 ${n} 次（豆制品除外） · ${strength}`
    case "prefer":
      return `优先安排${categoryLabel(tpl.category)}`
    case "avoid":
      return `尽量避开${categoryLabel(tpl.category)}`
    default:
      return ""
  }
}

export function parseTemplate(rule: MenuRule): MenuRuleTemplate | null {
  const raw = rule.template?.trim()
  if (!raw) return null
  try {
    const parsed = JSON.parse(raw) as MenuRuleTemplate
    if (!parsed || typeof parsed !== "object") return null
    if (!TYPE_OPTIONS.some((option) => option.key === parsed.type)) return null
    return parsed
  } catch {
    return null
  }
}

// Updates only the template; the backend overwrites expression/kind/severity
// from it on save, so the stale expression here is harmless.
export function withTemplate(rule: MenuRule, tpl: MenuRuleTemplate): MenuRule {
  return { ...rule, template: JSON.stringify(tpl) }
}

File: backend/internal/services/menu_rule_templates.go
package services

import (
	"encoding/json"
	"fmt"
	"strings"

	"ninimenu/internal/models"
)

// MenuRuleTemplate is the sentence-builder representation of a rule. The
// rendered expression stays the execution truth; the template only drives
// rendering and the friendly UI.
type MenuRuleTemplate struct {
	Type     string `json:"type"`     // limit | no_repeat | prefer | avoid
	Scope    string `json:"scope"`    // meal | day | week (limit / no_repeat)
	Category string `json:"category"` // egg/cold/staple/soup/spicy/heavy_spicy/deep_fry/slow/protein:K/ingredient:W
	N        int    `json:"n"`        // limit threshold / no_repeat overlap threshold
	Points   int    `json:"points"`   // score magnitude
	Strength string `json:"strength"` // must | prefer
}

type renderedMenuRule struct {
	Expression string
	RuleKind   string
	Severity   string
	Relaxable  bool
	Scope      string
}

// ApplyMenuRuleTemplate renders rule.Template (when present) into the
// executable fields, overwriting Expression/RuleKind/Severity/Relaxable/Scope.
func ApplyMenuRuleTemplate(rule *models.MenuRule) error {
	raw := strings.TrimSpace(rule.Template)
	if raw == "" {
		return nil
	}
	var template MenuRuleTemplate
	if err := json.Unmarshal([]byte(raw), &template); err != nil {
		return fmt.Errorf("规则模板格式无效: %v", err)
	}
	rendered, err := template.render()
	if err != nil {
		return err
	}
	rule.Expression = rendered.Expression
	rule.RuleKind = rendered.RuleKind
	rule.Severity = rendered.Severity
	rule.Relaxable = rendered.Relaxable
	rule.Scope = rendered.Scope
	return nil
}

func (t MenuRuleTemplate) render() (renderedMenuRule, error) {
	t.Type = strings.ToLower(strings.TrimSpace(t.Type))
	t.Strength = strings.ToLower(strings.TrimSpace(t.Strength))
	if t.Strength == "" {
		t.Strength = "prefer"
	}
	if t.Strength != "must" && t.Strength != "prefer" {
		return renderedMenuRule{}, fmt.Errorf("规则模板强度无效: %s", t.Strength)
	}
	t.Scope = strings.ToLower(strings.TrimSpace(t.Scope))
	if t.Scope == "" {
		t.Scope = "meal"
	}
	if t.Scope != "meal" && t.Scope != "day" && t.Scope != "week" {
		return renderedMenuRule{}, fmt.Errorf("规则模板范围无效: %s", t.Scope)
	}
	t.Category = strings.TrimSpace(t.Category)
	t.N = clampInt(t.N, 0, 10)
	if t.N == 0 {
		t.N = 1
	}
	t.Points = clampInt(t.Points, 0, 100)
	if t.Points == 0 {
		t.Points = defaultTemplatePoints(t.Type)
	}

	switch t.Type {
	case "limit":
		return t.renderLimit()
	case "no_repeat":
		return t.renderNoRepeat()
	case "prefer", "avoid":
		return t.renderPreference()
	default:
		return renderedMenuRule{}, fmt.Errorf("规则模板类型无效: %s", t.Type)
	}
}

func defaultTemplatePoints(templateType string) int {
	switch templateType {
	case "no_repeat":
		return 14
	case "prefer":
		return 20
	default: // limit, avoid
		return 15
	}
}

func (t MenuRuleTemplate) renderLimit() (renderedMenuRule, error) {
	predicate, err := menuRuleCategoryPredicate(t.Category)
	if err != nil {
		return renderedMenuRule{}, err
	}
	count, err := menuRuleCategoryCount(t.Category, t.Scope)
	if err != nil {
		return renderedMenuRule{}, err
	}
	if t.Strength == "must" {
		return renderedMenuRule{
			Expression: fmt.Sprintf("!(%s) || %s < %d", predicate, count, t.N),
			RuleKind:   menuRuleKindConstraint,
			Severity:   "hard",
			Relaxable:  false,
			Scope:      t.Scope,
		}, nil
	}
	return renderedMenuRule{
		Expression: fmt.Sprintf("%s && %s >= %d ? -%d : 0", predicate, count, t.N, t.Points),
		RuleKind:   menuRuleKindScore,
		Severity:   "soft",
		Relaxable:  true,
		Scope:      t.Scope,
	}, nil
}

func (t MenuRuleTemplate) renderNoRepeat() (renderedMenuRule, error) {
	overlap := fmt.Sprintf(`countOverlap%s("protein_sources", candidate.protein_sources)`, templateScopeSuffix(t.Scope))
	if t.Strength == "must" {
		return renderedMenuRule{
			Expression: fmt.Sprintf(`%s < %d || hasOnly(candidate.protein_sources, "soy")`, overlap, t.N),
			RuleKind:   menuRuleKindConstraint,
			Severity:   "hard",
			Relaxable:  false,
			Scope:      t.Scope,
		}, nil
	}
	return renderedMenuRule{
		Expression: fmt.Sprintf(`%s >= %d && !hasOnly(candidate.protein_sources, "soy") ? -%d : 0`, overlap, t.N, t.Points),
		RuleKind:   menuRuleKindScore,
		Severity:   "soft",
		Relaxable:  true,
		Scope:      t.Scope,
	}, nil
}

func (t MenuRuleTemplate) renderPreference() (renderedMenuRule, error) {
	predicate, err := menuRuleCategoryPredicate(t.Category)
	if err != nil {
		return renderedMenuRule{}, err
	}
	points := t.Points
	if t.Type == "avoid" {
		points = -points
	}
	return renderedMenuRule{
		Expression: fmt.Sprintf("%s ? %d : 0", predicate, points),
		RuleKind:   menuRuleKindScore,
		Severity:   "soft",
		Relaxable:  true,
		Scope:      "candidate",
	}, nil
}

func templateScopeSuffix(scope string) string {
	switch scope {
	case "day":
		return "Day"
	case "week":
		return "Week"
	default:
		return "Meal"
	}
}

func menuRuleCategoryPredicate(category string) (string, error) {
	switch {
	case category == "egg":
		return `has(candidate.protein_sources, "egg")`, nil
	case category == "cold":
		return `candidate.serving_temperature == "cold"`, nil
	case category == "staple":
		return `candidate.dish_role == "staple"`, nil
	case category == "soup":
		return `candidate.dish_role == "soup"`, nil
	case category == "spicy":
		return `candidate.spice_level >= 1`, nil
	case category == "heavy_spicy":
		return `candidate.spice_level >= 2 && candidate.richness_level >= 2`, nil
	case category == "deep_fry":
		return `has(candidate.cooking_methods, "deep_fry")`, nil
	case category == "slow":
		return `candidate.cook_time > 45`, nil
	case category == "non_favorite":
		return `!candidate.favorite`, nil
	case category == "unfamiliar_category":
		return `candidate.category != "" && candidate.category_favorites == 0`, nil
	case category == "weekday_slow_soup":
		return `!is_weekend && candidate.dish_role == "soup" && (candidate.cook_time > 45 || candidate.difficulty == "hard")`, nil
	case category == "slow_soup":
		return `candidate.dish_role == "soup" && candidate.cook_time > 45`, nil
	case category == "soup_ingredient_repeat":
		return `candidate.dish_role == "soup" && countOverlapPrevSoup("ingredients", candidate.ingredients) > 0`, nil
	case strings.HasPrefix(category, "protein:"):
		key, err := templateProteinKey(category)
		if err != nil {
			return "", err
		}
		return fmt.Sprintf(`has(candidate.protein_sources, %q)`, key), nil
	case strings.HasPrefix(category, "ingredient:"):
		word, err := templateIngredientWord(category)
		if err != nil {
			return "", err
		}
		return fmt.Sprintf(`(candidate.name contains %q || any(candidate.ingredients, # contains %q))`, word, word), nil
	default:
		return "", fmt.Errorf("规则模板类别无效: %s", category)
	}
}

func menuRuleCategoryCount(category string, scope string) (string, error) {
	suffix := templateScopeSuffix(scope)
	switch {
	case category == "egg":
		return fmt.Sprintf(`count%s("protein_sources", "egg")`, suffix), nil
	case category == "cold":
		return fmt.Sprintf(`count%s("serving_temperature", "cold")`, suffix), nil
	case category == "staple":
		return fmt.Sprintf(`count%s("dish_role", "staple")`, suffix), nil
	case category == "soup":
		return fmt.Sprintf(`count%s("dish_role", "soup")`, suffix), nil
	case category == "spicy":
		return fmt.Sprintf(`len(filter(%s, .spice_level >= 1))`, scope), nil
	case category == "heavy_spicy":
		return fmt.Sprintf(`count%s("heavy_spicy", "true")`, suffix), nil
	case category == "deep_fry":
		return fmt.Sprintf(`count%s("cooking_methods", "deep_fry")`, suffix), nil
	case category == "slow":
		return fmt.Sprintf(`len(filter(%s, .cook_time > 45))`, scope), nil
	case category == "slow_soup":
		return fmt.Sprintf(`len(filter(%s, .dish_role == "soup" && .cook_time > 45))`, scope), nil
	case strings.HasPrefix(category, "protein:"):
		key, err := templateProteinKey(category)
		if err != nil {
			return "", err
		}
		return fmt.Sprintf(`count%s("protein_sources", %q)`, suffix, key), nil
	case strings.HasPrefix(category, "ingredient:"):
		// v1 limitation: scope counting matches dish names only; the
		// candidate predicate still checks ingredients.
		word, err := templateIngredientWord(category)
		if err != nil {
			return "", err
		}
		return fmt.Sprintf(`len(filter(%s, .name contains %q))`, scope, word), nil
	default:
		return "", fmt.Errorf("规则模板类别无效: %s", category)
	}
}

func templateProteinKey(category string) (string, error) {
	key := strings.ToLower(strings.TrimSpace(strings.TrimPrefix(category, "protein:")))
	if !weekPlanProteinKeys[key] {
		return "", fmt.Errorf("规则模板蛋白类别无效: %s", category)
	}
	return key, nil
}

func templateIngredientWord(category string) (string, error) {
	word := strings.TrimSpace(strings.TrimPrefix(category, "ingredient:"))
	if word == "" {
		return "", fmt.Errorf("规则模板食材关键词不能为空")
	}
	if strings.ContainsAny(word, "\"\\`") {
		return "", fmt.Errorf("规则模板食材关键词包含非法字符")
	}
	return word, nil
}
